dev/src/dev/rewrite_defendpoint_1.clj

```diff
@@ -0,0 +1,422 @@
+
+;;;
+;;; Add metadata map if needed
+;;;
+
+(mu/defn- add-metadata-map :- ::zloc
+  [parsed      :- ::parsed
+   defendpoint :- ::zloc]
+  (if (= (n/tag (:method parsed)) :meta)
+    (let [metadata  (first (n/children (:method parsed)))
+          node      (if (= (n/tag metadata) :token)
+                      ;; TODO -- what about multiple tokens? Does this work?
+                      (n/map-node [metadata (n/whitespace-node " ") (n/token-node true)])
+                      metadata)]
+      (as-> defendpoint defendpoint
+        (z/insert-left (find-argslist defendpoint) node)
+        (z/insert-left (find-argslist defendpoint) (n/newline-node "\n"))))
+    defendpoint))
+
+;;;
+;;; rewrite body
+;;;
+
+(mu/defn- find-body :- ::zloc
+  [defendpoint :- ::zloc]
+  (z/right (find-argslist defendpoint)))
+
+(mu/defn- rewrite-body :- ::zloc
+  [_parsed     :- ::parsed
+   defendpoint :- ::zloc]
+  (let [body (find-body defendpoint)]
+    ;; if the first form is a map but there are more forms after, it's the schema map and we can remove it.
+    (if (and (= (z/tag body) :map)
+             (z/right body))
+      (z/remove body)
+      body)))
+
+;;;
+;;; rewrite entire form
+;;;
+
+(mu/defn- rewrite-defendpoint* :- ::zloc
+  [defendpoint :- ::zloc]
+  (let [parsed (parse-defendpoint-1-args defendpoint)]
+    (reduce
+     (fn [defendpoint f]
+       (z/subedit-node defendpoint (partial f parsed)))
+     defendpoint
+     [rewrite-defendpoint-symbol
+      rewrite-method
+      add-metadata-map
+      rewrite-args
+      rewrite-body])))
+
+(mu/defn- defendpoint-node? [node :- ::node]
+  (and
+   (= (n/tag node) :list)
+   (let [first-child (first (n/children node))]
+     (and
+      (n/symbol-node? first-child)
+      (= (n/sexpr first-child) 'api/defendpoint)))))
+
+(mu/defn- kondo-ignore-node? [node :- ::node]
+  (and
+   (= (n/tag node) :uneval)
+   (= (n/sexpr (first (n/children node)))
+      {:clj-kondo/ignore [:deprecated-var]})))
+
+(mu/defn- remove-preceding-kondo-ignore-node :- [:maybe ::zloc]
+  [defendpoint :- ::zloc]
+  (when-let [previous (z/left defendpoint)]
+    (when (some-> previous z/node kondo-ignore-node?)
+      (-> (z/subedit-node
+           (z/up defendpoint) ; move up to top-level <forms> node so the kondo ignore form is subeditable
+           (fn [_]
+             (z/remove previous)))
+          ; move back to first child
+          z/down))))
+
+(mu/defn- rewrite-defendpoint :- ::zloc
+  [defendpoint :- ::zloc]
+  (try
+    (-> defendpoint
+        rewrite-defendpoint*
+        remove-preceding-kondo-ignore-node)
+    (catch Throwable e
+      (println "Error rewriting defendpoint:" e)
+      defendpoint)))
+
+(mu/defn- rewrite-namespace :- ::zloc
+  [zloc :- ::zloc]
+  (if-let [defendpoint (z/find-next zloc #(defendpoint-node? (z/node %)))]
+    (let [zloc' (rewrite-defendpoint defendpoint)]
+      (recur zloc'))
+    zloc))
+
+;; TODO -- require needs to include api.macros.
```

Comment:
I didn't make this smart enough to add the require for `metabase.api.macros` to the `ns` form... yet. Probably not too hard tho TBH.